roachtest: tpccbench/nodes=9/cpu=4/chaos/partition failed

[(roachtest).tpccbench/nodes=9/cpu=4/chaos/partition failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1708438&tab=buildLog) on [master@3c408a47707a15d4b5c92031c4eafcfa643d307c](https://github.com/cockroachdb/cockroach/commits/3c408a47707a15d4b5c92031c4eafcfa643d307c):

```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/20200124-1708438/tpccbench/nodes=9/cpu=4/chaos/partition/run_1
	cluster.go:2315,tpcc.go:720,tpcc.go:561,test_runner.go:719: error with attached stack trace:
		    main.(*monitor).WaitE
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:2306
		    main.(*monitor).Wait
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:2314
		    main.runTPCCBench
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tpcc.go:720
		    main.registerTPCCBenchSpec.func1
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tpcc.go:561
		    main.(*testRunner).runTest.func2
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/test_runner.go:719
		    runtime.goexit
		    	/usr/local/go/src/runtime/asm_amd64.s:1357
		  - monitor failure:
		  - unexpected node event: 9: dead

```

<details><summary>More</summary><p>


Artifacts: [/tpccbench/nodes=9/cpu=4/chaos/partition](https://teamcity.cockroachdb.com/viewLog.html?buildId=1708438&tab=artifacts#/tpccbench/nodes=9/cpu=4/chaos/partition)
Related:
- #44293 roachtest: tpccbench/nodes=9/cpu=4/chaos/partition failed  [C-test-failure](https://api.github.com/repos/cockroachdb/cockroach/labels/C-test-failure) [O-roachtest](https://api.github.com/repos/cockroachdb/cockroach/labels/O-roachtest) [O-robot](https://api.github.com/repos/cockroachdb/cockroach/labels/O-robot) [branch-release-2.1](https://api.github.com/repos/cockroachdb/cockroach/labels/branch-release-2.1) [release-blocker](https://api.github.com/repos/cockroachdb/cockroach/labels/release-blocker)

- #44287 roachtest: tpccbench/nodes=9/cpu=4/chaos/partition failed  [C-test-failure](https://api.github.com/repos/cockroachdb/cockroach/labels/C-test-failure) [O-roachtest](https://api.github.com/repos/cockroachdb/cockroach/labels/O-roachtest) [O-robot](https://api.github.com/repos/cockroachdb/cockroach/labels/O-robot) [branch-release-19.1](https://api.github.com/repos/cockroachdb/cockroach/labels/branch-release-19.1) [release-blocker](https://api.github.com/repos/cockroachdb/cockroach/labels/release-blocker)

- #44259 roachtest: tpccbench/nodes=9/cpu=4/chaos/partition failed  [C-test-failure](https://api.github.com/repos/cockroachdb/cockroach/labels/C-test-failure) [O-roachtest](https://api.github.com/repos/cockroachdb/cockroach/labels/O-roachtest) [O-robot](https://api.github.com/repos/cockroachdb/cockroach/labels/O-robot) [branch-provisional_202001221958_v2.1.11](https://api.github.com/repos/cockroachdb/cockroach/labels/branch-provisional_202001221958_v2.1.11) [release-blocker](https://api.github.com/repos/cockroachdb/cockroach/labels/release-blocker)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Atpccbench%2Fnodes%3D9%2Fcpu%3D4%2Fchaos%2Fpartition.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/storage/store_create_replica.go
+++ b/pkg/storage/store_create_replica.go
@@ -85,18 +85,18 @@ func (s *Store) tryGetOrCreateReplica(
 		repl.raftMu.Lock() // not unlocked on success
 		repl.mu.Lock()
 
-		// Drop messages from replicas we know to be too old.
-		if fromReplicaIsTooOld(repl, creatingReplica) {
-			repl.mu.Unlock()
-			repl.raftMu.Unlock()
-			return nil, false, roachpb.NewReplicaTooOldError(creatingReplica.ReplicaID)
-		}
-
 		// The current replica is removed, go back around.
 		if repl.mu.destroyStatus.Removed() {
 			repl.mu.Unlock()
 			repl.raftMu.Unlock()
 			return nil, false, errRetry
+		}
+
+		// Drop messages from replicas we know to be too old.
+		if fromReplicaIsTooOld(repl, creatingReplica) {
+			repl.mu.Unlock()
+			repl.raftMu.Unlock()
+			return nil, false, roachpb.NewReplicaTooOldError(creatingReplica.ReplicaID)
 		}
 
 		toTooOld := toReplicaIsTooOld(repl, replicaID)
@@ -187,6 +187,22 @@ func (s *Store) tryGetOrCreateReplica(
 	// Store.mu to maintain lock ordering invariant.
 	repl.mu.Lock()
 	repl.mu.tombstoneMinReplicaID = tombstone.NextReplicaID
+	uninitializedDesc := &roachpb.RangeDescriptor{
+		RangeID: rangeID,
+		// NB: other fields are unknown; need to populate them from
+		// snapshot.
+	}
+	// NB: A Replica should never be in the store's replicas map with a nil
+	// descriptor. Assign it directly here. In the case that the Replica should
+	// exist (which we confirm with another check of the Tombstone below), we'll
+	// re-initialize the replica with the same uninitializedDesc.
+	//
+	// During short window between here and call to s.unlinkReplicaByRangeIDLocked()
+	// in the failure branch below, the Replica used to have a nil descriptor and
+	// was present in the map. While it was the case that the destroy status had
+	// been set, not every code path which inspects the descriptor checks the
+	// destroy status.
+	repl.mu.state.Desc = uninitializedDesc
 	// Add the range to range map, but not replicasByKey since the range's start
 	// key is unknown. The range will be added to replicasByKey later when a
 	// snapshot is applied. After unlocking Store.mu above, another goroutine
@@ -224,16 +240,10 @@ func (s *Store) tryGetOrCreateReplica(
 		} else if hs.Commit != 0 {
 			log.Fatalf(ctx, "found non-zero HardState.Commit on uninitialized replica %s. HS=%+v", repl, hs)
 		}
-
-		desc := &roachpb.RangeDescriptor{
-			RangeID: rangeID,
-			// NB: other fields are unknown; need to populate them from
-			// snapshot.
-		}
-		return repl.initRaftMuLockedReplicaMuLocked(desc, replicaID)
+		return repl.initRaftMuLockedReplicaMuLocked(uninitializedDesc, replicaID)
 	}(); err != nil {
 		// Mark the replica as destroyed and remove it from the replicas maps to
-		// ensure nobody tries to use it
+		// ensure nobody tries to use it.
 		repl.mu.destroyStatus.Set(errors.Wrapf(err, "%s: failed to initialize", repl), destroyReasonRemoved)
 		repl.mu.Unlock()
 		s.mu.Lock()

--- a/pkg/storage/store_test.go
+++ b/pkg/storage/store_test.go
@@ -18,6 +18,7 @@ import (
 	"math/rand"
 	"reflect"
 	"sort"
+	"sync"
 	"sync/atomic"
 	"testing"
 	"time"
@@ -2690,6 +2691,35 @@ func TestStoreGCThreshold(t *testing.T) {
 	}
 
 	assertThreshold(threshold)
+}
+
+// TestRaceOnTryGetOrCreateReplicas exercises a case where a race between
+// different raft messages addressed to different replica IDs could lead to
+// a nil pointer panic.
+func TestRaceOnTryGetOrCreateReplicas(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	tc := testContext{}
+	stopper := stop.NewStopper()
+	ctx := context.Background()
+	defer stopper.Stop(ctx)
+	tc.Start(t, stopper)
+	s := tc.store
+	var wg sync.WaitGroup
+	for i := 3; i < 100; i++ {
+		wg.Add(1)
+		go func(rid roachpb.ReplicaID) {
+			defer wg.Done()
+			r, _, _ := s.getOrCreateReplica(ctx, 42, rid, &roachpb.ReplicaDescriptor{
+				NodeID:    2,
+				StoreID:   2,
+				ReplicaID: 2,
+			}, false)
+			if r != nil {
+				r.raftMu.Unlock()
+			}
+		}(roachpb.ReplicaID(i))
+	}
+	wg.Wait()
 }
 
 func TestStoreRangePlaceholders(t *testing.T) {

